plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/schema/confluent/AbstractConfluentRowEncoder.java

```diff
@@ -0,0 +1,246 @@
+            Type elementType = arrayType.getElementType();
+            Block arrayBlock = block.getObject(position, Block.class);
+            List<Object> list = new ArrayList<>(arrayBlock.getPositionCount());
+            Schema elementAvroType = fieldSchema.getElementType();
+            for (int i = 0; i < arrayBlock.getPositionCount(); i++) {
+                list.add(getValue(elementType, arrayBlock, i, elementAvroType));
+            }
+            return Collections.unmodifiableList(list);
+        }
+        else if (type instanceof MapType) {
+            MapType mapType = (MapType) type;
+            Block mapBlock = block.getObject(position, Block.class);
+            Schema valueSchema = fieldSchema.getValueType();
+            Type valueType = mapType.getValueType();
+            Map<String, Object> map = new HashMap<>();
+            for (int index = 0; index < mapBlock.getPositionCount(); index += 2) {
+                String key = VARCHAR.getSlice(mapBlock, index).toStringUtf8();
+                Object value = getValue(valueType, mapBlock, index + 1, valueSchema);
+                map.put(key, value);
+            }
+            return Collections.unmodifiableMap(map);
+        }
+        else if (type instanceof RowType) {
+            RowType rowType = (RowType) type;
+            GenericRecordBuilder recordBuilder = new GenericRecordBuilder(fieldSchema);
+            // If the avro schema is a record with no fields and the empty field strategy is ADD_DUMMY, ignore the dummy value being inserted.
+            if (!isEmptyStruct(fieldSchema, rowType)) {
+                checkState(rowType.getFields().size() == fieldSchema.getFields().size(), "Mismatch in number of fields in row and schema");
+                Block rowBlock = block.getObject(position, Block.class);
+                for (int index = 0; index < rowType.getFields().size(); index++) {
+                    RowType.Field field = rowType.getFields().get(index);
+                    Schema.Field avroField = fieldSchema.getFields().get(index);
+                    Schema avroFieldSchema = avroField.schema();
+                    Object value = getValue(field.getType(), rowBlock, index, avroFieldSchema);
+                    recordBuilder.set(avroField, value);
+                }
+            }
+            return recordBuilder.build();
+        }
+        else {
+            throw new UnsupportedOperationException(format("Unsupported type '%s'", type));
+        }
+    }
+
+    @VisibleForTesting
+    public static Schema extractBaseType(Schema schema)
+    {
+        requireNonNull(schema, "schema is null");
+        if (!schema.isUnion()) {
+            return schema;
+        }
+        Optional<Schema> baseSchema = schema.getTypes().stream()
+                .filter(fieldSchema -> fieldSchema.getType() != Schema.Type.NULL)
+                .map(nonNullSchema -> {
+                    if (nonNullSchema.isUnion()) {
```

Comment:
what if it's a union type of `string` and `Foo`? it seems this would arbitrarily return `string` or `Foo`.

Since we don't support such unions we should disallow it here.

Reply:
Good point! Now it will only support UNION(NULL, <TYPE>) where TYPE can also be UNION(NULL, <TYPE>) but there must only be one type.